File: 03/exercise_03.py
import sys
sys.path.append("..")
import helpers


LETTERS = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
LETTER_SCORE = {char: idx + 1 for idx, char in enumerate(LETTERS)}


def part_one(input_filename):
    input = helpers.parse_input(input_filename)
    score = 0
    for line in input:
        midpoint = len(line) // 2
        first, second = set(line[0:midpoint]), set(line[midpoint:])
        for letter in first.intersection(second):  # Only 1 element in the intersection
            score += LETTER_SCORE[letter]
    return score


def part_two(input_filename):
    input = helpers.parse_input(input_filename)
    ticker = 0
    working_set = None
    score = 0
    for line in input:
        working_set = set(line) if not working_set else working_set.intersection(set(line))
        ticker += 1
        if ticker == 3:
            letter = next(iter(working_set))  # Only 1 element in the set
            score += LETTER_SCORE[letter]
            working_set = None
            ticker = 0
    return score


if __name__ == "__main__":
    print("*** PART ONE ***\n")
    print(f"Test result = {part_one('inputtest.txt')}\n")
    print(f"REAL RESULT = {part_one('input.txt')}\n\n")
    print("*** PART TWO ***\n")
    print(f"Test result = {part_two('inputtest.txt')}\n")
    print(f"REAL RESULT = {part_two('input.txt')}")

File: 03/exercise_03.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

const LETTERS = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"

func make_map(string) map[rune]int {
	score_map := make(map[rune]int)
	for idx, char := range LETTERS {
		score_map[char] = idx + 1
	}
	fmt.Println(score_map)
	return score_map
}

var LETTER_SCORES = make_map(LETTERS)

func part1(input_filename string) string {

	score := 0
	file, _ := os.Open(input_filename)
	defer file.Close()
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		text := scanner.Text()
		midpoint := len(text) / 2
		part1, part2 := text[0:midpoint], text[midpoint:]
		for _, char := range part1 {
			if strings.ContainsRune(part2, char) {
				score += LETTER_SCORES[char]
				break
			}
		}
	}
	return strconv.Itoa(score)
}

func part2(input_filename string) string {

	score, ticker := 0, 0
	working_string := ""
	file, _ := os.Open(input_filename)
	defer file.Close()
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		text := scanner.Text()
		ticker += 1
		newstring := ""
		if working_string == "" {
			working_string = text
		} else {
			for _, char := range working_string {
				if strings.ContainsRune(text, char) {
					newstring += string(char)
				}
			}
			working_string = newstring
		}
		if ticker%3 == 0 {
			score += LETTER_SCORES[rune(working_string[0])]
			working_string = ""
		}
	}
	return strconv.Itoa(score)
}

func main() {

	fmt.Println(part1("inputtest.txt"))
	fmt.Println(part1("input.txt"))
	fmt.Println(part2("inputtest.txt"))
	fmt.Println(part2("input.txt"))
}
